src/repositories/request-repository.ts

```diff
@@ -152,7 +157,14 @@ const findRequestsToAnchorForStreams = (
   streamIds: string[],
   now: Date
 ): Promise<Array<Request>> =>
-  findRequestsToAnchor(connection, now).whereIn('streamId', streamIds).orderBy('createdAt', 'asc')
+  findRequestsToAnchor(connection, now)
+    .whereIn('streamId', streamIds)
+    .orderByRaw(
+      `array_position(ARRAY[${streamIds.map(
+        (streamId) => `'${streamId}'`
+      )}], stream_id), created_at ASC`
+    )
+    .limit(POSTGRES_PARAMETERIZED_QUERY_LIMIT)
```

Comment:
ugh, unfortunately this means that if the postgres request limit falls in the middle of a group of requests for the same stream, you could wind up anchoring an older request instead of a newer one and cause the newer commits to wind up getting rolled back when the anchor commit supercedes them.

I think if you make the query sort first by stream_id and second by created_at in descending order, then even if half of the requests get cut off for a stream, at least you'll consider the newer requests first, and the next batch will pick up the older requests and hopefully see that they were actually already anchored implicitly

Reply:
> ugh, unfortunately this means that if the postgres request limit falls in the middle of a group of requests for the same stream, you could wind up anchoring an older request instead of a newer one and cause the newer commits to wind up getting rolled back when the anchor commit supercedes them.

When a stream gets anchored, the tip of the stream is anchored, regardless of the commits in the requests. So we will not be anchoring at an older commit

When the next batch with the more recent requests gets anchored, this stream will be already anchored and will not be included in the merkle tree (unless a new commit has come in).

The problem I can see this exasperating is the zktruth error (https://linear.app/3boxlabs/issue/CDB-1960/request-status-for-non-tip-commit-throwing) as the anchor commit stored in the postgres will reference an older request. But we already have this problem for several scenarios, and when a fix is implemented this problem will go away.

We can sort the created_at in descending order BUT it may cause this scenario:
MyStream has requests for Commit1 <-- Commit2 <-- ... <-- Commit10000
The request for Commit1 is expiring. 
We create Batch1 and anchor Commit5000 <-- ... <-- Commit1000. We mark all those requests as COMPLETED
We create Batch2 immediately because Commit1 is expiring. Batch 2 is not full but had to run because the request for Commit1 is expiring.

All of this again is quite unlikely. 

And then when ceramic w/o a CAS node should get rid of this problem =)

Reply:
yeah, although anchoring the tip depends on the Ceramic node correctly having the current tip.  Having the request in the batch means the CAS will ensure that the Ceramic node has considered the request's commit if it doesn't otherwise show up in the stream log.

But this is a rare enough thing to have happen in the first place and that's another edge case on top, so I guess it doesn't really matter that much.